sstables/compaction_manager.cc

```diff
@@ -599,19 +599,30 @@ void compaction_manager::submit(column_family* cf) {
                 _stats.active_tasks--;
                 task->compaction_running = false;
 
-                if (!can_proceed(task)) {
-                    maybe_stop_on_error(std::move(f), stop_iteration::yes);
+                try {
+                    f.get();
+                    _stats.completed_tasks++;
+                } catch (sstables::compaction_stop_exception& e) {
+                    cmlog.info("compaction info: {}: stopping", e.what());
+                    _stats.errors++;
+                    // if regular compaction stopped on user's request, let's abort the task and allow next attempt to retry it.
                     return make_ready_future<stop_iteration>(stop_iteration::yes);
-                }
-                if (maybe_stop_on_error(std::move(f))) {
+                } catch (storage_io_error& e) {
+                    cmlog.error("compaction failed due to storage io error: {}: stopping", e.what());
                     _stats.errors++;
-                    _stats.pending_tasks++;
-                    return put_task_to_sleep(task).then([] {
+                    do_stop();
+                    return make_ready_future<stop_iteration>(stop_iteration::yes);
+                } catch (...) {
+                    _stats.errors++;
+                    if (!can_proceed(task)) {
+                        cmlog.error("compaction failed {}: stopping", std::current_exception());
+                        return make_ready_future<stop_iteration>(stop_iteration::yes);
+                    }
+                    cmlog.error("compaction failed {}: retrying", std::current_exception());
+                    return put_task_to_sleep(task).then([this] {
```

Comment:
But now the code is duplicated with maybe_stop_on_error.

(agree moving a future into it is ugly, but so is the duplication).